src/realm/index_string.cpp

```diff
@@ -197,6 +197,39 @@ size_t IndexArray::from_list<index_FindAll_nocopy>(StringData value, IntegerColu
     return size_t(FindRes_column);
 }
 
+
+template <>
+size_t IndexArray::from_list<index_FindAll_ins>(StringData value, IntegerColumn& result, InternalFindResult& /*result_ref*/,
+                                                const IntegerColumn& rows, ColumnBase* column) const
+{
+    // The buffer is needed when for when this is an integer index.
+    StringIndex::StringConversionBuffer buffer;
+
+    // optimization for the most common case, where all the strings under a given subindex are equal
+    StringData first_str = column->get_index_data(to_size_t(*rows.cbegin()), buffer);
+    StringData last_str = column->get_index_data(to_size_t(*(rows.cend() - 1)), buffer);
+    if (first_str == last_str) {
+        for (IntegerColumn::const_iterator it = rows.cbegin(); it != rows.cend(); ++it) {
+            const size_t row_ndx = to_size_t(*it);
+            result.add(row_ndx);
+        }
+        return size_t(FindRes_column); // is ignored
+    }
+
+    // special case for very long strings, where they might have a common prefix and end up in the
+    // same subindex column, but still not be identical
+    auto upper_value = case_map(value, true);
+    for (IntegerColumn::const_iterator it = rows.cbegin(); it != rows.cend(); ++it) {
+        const size_t row_ndx = to_size_t(*it);
+        StringData str = column->get_index_data(row_ndx, buffer);
+        auto upper_str = case_map(str, true);
+        if (upper_str == upper_value)
+            result.add(row_ndx);
+    }
```

Comment:
True. It might be even better to do a binary search. However, this path is for strings with common prefixes of at least 200 bytes, so I think we can defer optimising (& add complexity) this until (if ever) it becomes necessary.